c10/core/TensorImpl.h

```diff
@@ -175,6 +175,109 @@ struct C10_API AutogradMetaFactoryRegisterer {
+//    holds the GIL for the interpreter tagged on the TensorImpl.  (If the
+//    TensorImpl is not tagged, it must first atomically claim its tag before it
+//    can validly write)
+
+// The PyInterpreter object itself is a class that contains some function
+// pointers for interacting with the interpreter.  For now this is just for
+// debugging, but if a Tensor can own a PyObject, the interpreter can be used to
+// free it.
+//
+// WARNING: This class has to be written very carefully, because it may be
+// possible for a Tensor to have a reference an interpreter corresponding to
+// a shared library that has ALREADY BEEN UNLOADED.  This makes blindly calling
+// virtual methods very dangerous, because the vtable may be garbage at that
+// point (on a good day, you might get "pure virtual method called").
+//
+// The idea to solve this problem is we always leak PyInterpreters (so they
+// always stay live even after dlclose), and disarm the "virtual methods" by
+// replacing them with function pointers that just no-op.  This can't be done
+// with a traditional C++ vtable, so we have to roll our own.
+//
+// NB: The downside with representing PyInterpreter tags as full objects is that
+// it takes an extra word on TensorImpl.  If tags were instead just integer
+// indices, on 64-bit architectures we could pack the tag and PyObject together
+// into a single atomic word.  On 32-bit architectures we could simply say that
+// only one Python interpreter is supported (erroring if a nontrivial
+// interpreter tag is attempted to be set).
+//
+// The difficulty with this scheme is we need to maintain an out-of-line table
+// to get at the PyInterpreters so that we can do virtual method calls on them,
+// and registration/deregistration to this table must be done in a thread safe
+// manner.  This can be easily done if the number of possible PyInterpreters is
+// small enough (e.g., 8-bit integer) by simply preallocating an array of
+// sufficient size to hold all possible interpreters.  Surely 128 threads is
+// more than enough for anyone!
+//
+// I didn't decide to do this technique at the moment, because the extra word
+// added by the PyInterpreter tag takes us to 24 words, which means that we
+// still fit inside three eight word cache lines.  If you need to penny pinch
+// another word consider doing this!
+
+struct PyInterpreter;
+C10_API std::string noop_name_fn(const PyInterpreter*);
+struct C10_API PyInterpreter {
+  using name_sig = std::string(const PyInterpreter*);
+
+  PyInterpreter(name_sig* name_fn) : name_fn_(name_fn) {}
+
+  // For debugging purposes only
+  name_sig* name_fn_;
+
+  std::string name() const {
+    return (*name_fn_)(this);
+  }
+
+  // Disarm this PyInterpreter, making all of its methods noops.
+  // This is not really thread safe, but the hope is that long lived tensors
+  // will die after a synchronization even from the dlclose, typically when
```

Comment:
can you elaborate on the hope and the synchronization?

Reply:
```
  // Disarm this PyInterpreter, making all of its methods noops.
  // Because the function pointers are raw pointers (not atomics),
  // a disarm() invocation that is concurrent with active destructors
  // is not thread safe and will trigger TSAN.  My hope is that this
  // situations doesn't ever actually happen; tensor destruction should
  // quiesce when a dlclose happens, and any long lived tensors whose
  // destructors would be disarmed here only begin the destruction process
  // on process shutdown (long after the dlclose has occurred).

```